crates/noirc_evaluator/src/ssa/block.rs

```diff
@@ -0,0 +1,195 @@
+use super::{code_gen::IRGenerator, node};
+use arena::Index;
+use std::collections::{HashMap, VecDeque};
+
+#[derive(PartialEq, Debug)]
+pub enum BlockType {
+    Normal,
+    ForJoin,
+}
+#[derive(Debug)]
+pub struct BasicBlock {
+    pub idx: arena::Index,
+    pub kind: BlockType,
+    pub dominator: Option<arena::Index>, //direct dominator
+    pub dominated: Vec<arena::Index>,    //dominated sons
+    pub predecessor: Vec<arena::Index>,  //for computing the dominator tree
+    pub left: Option<arena::Index>,      //sequential successor
+    pub right: Option<arena::Index>,     //jump successor
+    pub instructions: Vec<arena::Index>,
+    pub value_map: HashMap<arena::Index, arena::Index>, //for generating the ssa form
+}
+
+impl BasicBlock {
+    pub fn new(prev: arena::Index, kind: BlockType) -> BasicBlock {
+        BasicBlock {
+            idx: crate::ssa::code_gen::IRGenerator::dummy_id(),
+            predecessor: vec![prev],
+            left: None,
+            right: None,
+            instructions: Vec::new(),
+            value_map: HashMap::new(),
+            dominator: None,
+            dominated: Vec::new(),
+            kind,
+        }
+    }
+
+    pub fn get_current_value(&self, idx: arena::Index) -> Option<arena::Index> {
+        self.value_map.get(&idx).copied()
+    }
+
+    //When generating a new instance of a variable because of ssa, we update the value array
+    //to link the two variables
+    pub fn update_variable(&mut self, old_value: arena::Index, new_value: arena::Index) {
+        self.value_map.insert(old_value, new_value);
+    }
+
+    pub fn get_first_instruction(&self) -> arena::Index {
+        self.instructions[0]
+    }
+
+    pub fn is_join(&self) -> bool {
+        self.kind == BlockType::ForJoin
+    }
+}
+
+///////////
+
+pub fn create_first_block(igen: &mut IRGenerator) {
+    let first_block = BasicBlock::new(igen.dummy(), BlockType::Normal);
+    let new_idx = igen.blocks.insert(first_block);
+    let block2 = igen.blocks.get_mut(new_idx).unwrap(); //RIA..
+    block2.idx = new_idx;
+    igen.first_block = new_idx;
+    igen.current_block = new_idx;
+    igen.new_instruction(
+        igen.dummy(),
```

Comment:
Instead of using a dummy index, I wonder if we can use an Option::None

Reply:
It will make the code less readable because almost all instructions have two operands. 
For instructions having only one, normally I re-use the left for the right.
So nop should be the only one with dummy operands.

Reply:
Since the number of arguments is dependent on the instruction type, maybe they should be attached to the instruction type enum itself?